crates/neon/src/object/wrap.rs

```diff
@@ -0,0 +1,180 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl error::Error for WrapError {}
+
+impl<T> ResultExt<T> for Result<T, WrapError> {
+    fn or_throw<'cx, C>(self, cx: &mut C) -> NeonResult<T>
+    where
+        C: Context<'cx>,
+    {
+        match self {
+            Ok(v) => Ok(v),
+            Err(err) => cx.throw_error(err.to_string()),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum WrapErrorType {
+    AlreadyWrapped,
+    NotWrapped,
+    #[cfg(feature = "napi-8")]
+    ForeignType,
+}
+
+impl fmt::Display for WrapErrorType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::AlreadyWrapped => write!(f, "Object is already wrapped"),
+            Self::NotWrapped => write!(f, "Object is not wrapped"),
+            #[cfg(feature = "napi-8")]
+            Self::ForeignType => write!(f, "Object is wrapped by another addon"),
+        }
+    }
+}
+
+pub fn wrap<T, V>(cx: &mut Cx, o: Handle<V>, v: T) -> NeonResult<Result<(), WrapError>>
+where
+    T: Finalize + 'static,
+    V: Object,
+{
+    let env = cx.env().to_raw();
+    let o = o.to_local();
+    let v = Box::into_raw(Box::new(Box::new(v) as BoxAny));
+
+    // # Safety
+    // The `finalize` function will be called when the JavaScript object is garbage
+    // collected. The `data` pointer is guaranteed to be the same pointer passed when
+    // wrapping.
+    unsafe extern "C" fn finalize<T>(env: sys::Env, data: *mut c_void, _hint: *mut c_void)
+    where
+        T: Finalize + 'static,
+    {
+        let data = Box::from_raw(data.cast::<BoxAny>());
+        let data = *data.downcast::<T>().unwrap();
+        let env = Env::from(env);
+
+        Cx::with_context(env, move |mut cx| data.finalize(&mut cx));
+    }
+
+    // # Safety
+    // The `env` value was obtained from a valid `Cx` and the `o` handle has
+    // already been verified to be an object.
+    unsafe {
+        match sys::wrap(
+            env,
+            o,
+            v.cast(),
+            Some(finalize::<T>),
+            ptr::null_mut(),
+            ptr::null_mut(),
+        ) {
+            Err(sys::Status::InvalidArg) => {
+                // Wrap failed, we can safely free the value
+                let _ = Box::from_raw(v);
+
+                return Ok(Err(WrapError::already_wrapped()));
+            }
+            Err(sys::Status::PendingException) => {
+                // Wrap failed, we can safely free the value
+                let _ = Box::from_raw(v);
+
+                return Err(Throw::new());
+            }
+            // If an unexpected error occurs, we cannot safely free the value
+            // because `finalize` may be called later.
+            res => res.unwrap(),
+        }
+
+        #[cfg(feature = "napi-8")]
+        match sys::type_tag_object(env, o, &*crate::MODULE_TAG) {
+            Err(sys::Status::InvalidArg) => return Ok(Err(WrapError::foreign_type())),
```

Comment:
Should we attempt to `remove_wrap` before throwing this error? It seems bad to add the wrapping and then discover that we couldn't do the type tag.

Annoyingly this then means we have the possibility of an error case within the error case.

Or maybe we should do the type tag first, since if it fails it won't have mutated the object at all.